homeassistant/components/arcam_fmj/config_flow.py

```diff
@@ -1,12 +1,56 @@
 """Config flow to configure the Arcam FMJ component."""
-from operator import itemgetter
+import asyncio
+import logging
+from urllib.parse import urlparse
+
+import aiohttp
+from defusedxml import ElementTree
+import voluptuous as vol
 
 from homeassistant import config_entries
+from homeassistant.components.ssdp import ATTR_SSDP_LOCATION, ATTR_UPNP_UDN
 from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_UUID, DEFAULT_NAME, DEFAULT_PORT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _strip_uuid_prefix(data):
+    assert data.startswith("uuid:"), "Missing uuid prefix"
+    return data[5:].split("-")[4]
+
 
-from .const import DOMAIN
+async def _get_uniqueid_from_device_description(hass, url):
+    """Retrieve and extract unique id from ."""
+    try:
+        session = async_get_clientsession(hass)
+        async with session.get(url) as req:
+            req.raise_for_status()
+            xml = ElementTree.fromstring(await req.text())
 
-_GETKEY = itemgetter(CONF_HOST, CONF_PORT)
+        namespaces = {"d": "urn:schemas-upnp-org:device-1-0"}
+        udn = xml.findtext("d:device/d:UDN", None, namespaces)
+        assert udn, "Failed to get udn tag in xml"
+        return _strip_uuid_prefix(udn)
+
+    except (
+        aiohttp.ClientError,
+        asyncio.TimeoutError,
+        ElementTree.ParseError,
+        AssertionError,
+    ):
+        _LOGGER.exception("Unable to get device description from %s", url)
+
+    return None
+
+
+async def _get_uniqueid_from_host(hass, host):
+    """Try to deduce a unique id from a host based on ssdp/upnp."""
+    return await _get_uniqueid_from_device_description(
+        hass, f"http://{host}:8080/dd.xml"
+    )
```

Comment:
This is a bit of a hack. The receivers i know have this url format. That may or may not be true for other versions. In that case we'd have to do a directed SSDP query to get the url. But i think this is fine for now. Will fix should i get a report where it doesn't work.